Volumetric contact

galoompa opened the pull request:
Fixes issue #<issue_number>

### Brief summary of changes

### Testing I've completed

### Looking for feedback on...

### CHANGELOG.md (choose one)

- no need to update because...
- updated...

The Doxygen for this PR can be viewed at http://myosin.sourceforge.net/?C=N;O=D; click the folder whose name is this PR's number.

galoompa committed:
fix BodyDragForce equations

previously, only worked for even values of exponent

galoompa committed:
initial working volumetric contact

ellipsoid-plane contact equations were put in Examples\Plugins\ (based on structure of BodyDrag)

galoompa committed:
clean up comments

galoompa committed:
use sockets and improve constructor

The two frames are now implemented using sockets, by inheriting from TwoFrameLinker (similar to BushingForce).
A proper constructor was added (though more should be made available).

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+//=============================================================================
+// PROPERTIES
+//=============================================================================
+    // TODO: look for an osimsimulation library registered type, and attach it to that library instead of a separate one?
```
You can remove this comment. If you include "osimPluginDLL.h" above that means export as osimPluginDLL and that its registerTypes will contain your force. If you don't include it, the Force already includes "OpenSim/Simulation/osimSimulationDLL.h" and you only have to add your force to RegisterTypes_osimSimulation.cpp (e.g. `BushingForce` and other Forces in the Object list)

galoompa replied:
Which folder would it make sense to move this to? If I were to include it in osimSimulation, then it should probably not be a plugin example.

galoompa replied:
The files have been moved to Simulation/Model/
The other force components are also here; do you think this makes sense?

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+//=============================================================================
+    // TODO: look for an osimsimulation library registered type, and attach it to that library instead of a separate one?
+
+    OpenSim_DECLARE_PROPERTY(k_V, double,
```
We generally use more verbose property names in case the comment is missing from the model file. Here `volumetric_stiffness` or just `stiffness` would work.

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+//=============================================================================
+//=============================================================================
+/*
+ * A class template for writing a custom Force plugin. 
```
A description of your class including how it works and its default settings would be helpful.

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+/*
+ * A class template for writing a custom Force plugin. 
+ * Applies ...
+ * @author         
```
Don't be shy to list yourself as the author. Refer to other works where appropriate.

galoompa replied:
Where is it appropriate to list other works? Certainly it would be useful to, but I do not know how visible most of these are to users. What documentation do most users look at? The source code? the C++ API?

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+    VolumetricContactForce(const std::string& name,
+        const PhysicalFrame& ellipsoidFrame, const SimTK::Transform& transformInEllipsoidFrame,
+        const PhysicalFrame& planeFrame, const SimTK::Transform& transformInPlaneFrame,
+        const double& k_V,
```
Good to also use the same verbose variable names here so it is self documenting.

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+// METHODS
+//=============================================================================
+public:
+    // Default Constructor
```
It would be good to include what the default settings are.

galoompa replied:
For default constructors, what are parameters generally set to? Should I set them to something invalid, so ensure that the user enters values, or perhaps pick arbitrary values?

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+
+protected:
+    // TODO: is there any reason this should be implemented? It was in BodyDragForce
+    //virtual void connectToModel(Model& aModel);
```
`Component::connectToModel` is not virtual. It should be removed from `BodyDragForce`. If there is any additional connection related setup or checking that is required, then override `extendConnectToModel` (don't forget to call `Super::extendConnectToModel` first in your override).  Otherwise, sockets are handled as part of the machinery of the base `Component` and you don't have to do anything to test the validity of their connections.

galoompa replied:
Should I update BodyDragForce to fix some of these? Or shall I just focus on my volumetric contact work for this update?

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+#ifndef _OPENSIM_VOLUMETRIC_CONTACT_FORCE_PLUGIN_H_
+#define _OPENSIM_VOLUMETRIC_CONTACT_FORCE_PLUGIN_H_
+/* -------------------------------------------------------------------------- *
+ *                        VolumetricContactForce.h                            *
```
`VolumetricContactForce` might be a good base name for a family of elements especially if they will share the same (subset of) properties and sockets. Perhaps this class be called something like `EllipsoidHalfspaceVolumetricForce'?

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.cpp:
```diff
@@ -0,0 +1,289 @@
+
+    // Debuging info
+    // --------------
+    int deb = 0;
```
To make debugging info conditional you can use `Object::getDebugLevel()`.

galoompa replied:
I just removed this section of code - I was not making much use of it

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+namespace OpenSim {
+    // TODO: OSIMPLUGIN_API will change to OSEMSIMULATION_API
+class OSIMPLUGIN_API VolumetricContactForce 
+    : public TwoFrameLinker<Force, PhysicalFrame>
```
I don't think you want to use the `TwoFrameLinker` Mixin.  It  is a utility class for joining two frames useful for Joints, Constrainst and Bushing type Forces. If you are using methods of `TwoFrameLinker` in your calculations, then maybe it is OK, but it looked like you were just using the Frames. In this case, I would simply derive from `Force` (not the Mixin) and create two `Sockets<PhysicalFrame>` for `ellipsoid_frame` and  `halfspace_frame`.

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+//=============================================================================
+// Headers define the various property types that OpenSim objects can read 
+#include <string>
+#include "osimPluginDLL.h"
```
can remove when you make this class part of the osimSimulatuion library (i.e. use OSIMSIMULATION_API macro instead of OSIMPLUGIN_API)

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+                      SimTK::Vector_<SimTK::SpatialVec>& bodyForces, 
+                      SimTK::Vector& generalizedForces) const override;
+
+    /** This may be more difficult for volumetric... */
```
You can remove. The default implementation is to return 0.

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.cpp:
```diff
@@ -0,0 +1,289 @@
+    return get_mu_d() * tanh(4 * (v / vt)) + (get_mu_s() - get_mu_d()) * (v / vt) / pow(0.25 * pow((v / vt), 2) + 0.75, 2);
+}
+
+/** Potential energy function */
```
Not necessary to implement. It can be removed from here.

aseth1 commented on OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h:
```diff
@@ -0,0 +1,134 @@
+#ifndef _OPENSIM_VOLUMETRIC_CONTACT_FORCE_PLUGIN_H_
```
Update this to be _OPENSIM_<YOUR_CLASS_NAME> once you decide what you want to name it. _PLUGIN_ is unnecessary.

aseth1 reviewed (changes requested):
It is coming together. Mostly minor comments. The main things are the name of your class and deriving from `TwoFrameLinker`.  I did not review the math, which I assume you can verify using some benchmarks. Please include the VolumetricContactExample code and your test cases. If you add you class to the osimSimulation library, then you can add your test cases to testForces.

galoompa commented on OpenSim/Examples/Plugins/BodyDragExample/BodyDragForce.cpp:
```diff
@@ -145,7 +145,8 @@ void BodyDragForce::computeForce(const SimTK::State& s,
     // and the force vector itself to be in the body frame
```
I'm fairly certain that this is incorrect - the point should be in the body frame and the force in the ground frame. Shall I correct this as well? Or do you not care too much about the dragForce?

galoompa committed:
remove reference to TwoFrameLinker from volumetric contact

VolumetricContactForce was renamed to EllipsoidHalfSpaceVolumetricContactForce, and no longer inherits from TwoFrameLinker

galoompa committed:
Move volumetric contact into osimSimulation library

The ellipsoid-helfspace volumetric contact was moved out of a custom plugin (and associated code was deleted), and put in Simulation/Model

galoompa committed:
implement generateDecorations to show an ellipsoid

galoompa committed:
Add test cases for volumetric contact
